### PR TITLE
Source yapf file from project root

### PR DESCRIPTION
**Pull request by galaunay**
Copy of the PR #989 from @jonmoore.

Plus some documentation for the potential users of yapf style file.

### CODE REVIEW

**Comment by jorgenschaefer** [comment]
Looks good to me. I guess you can just always pass the directory there, reducing a lot of conditionals, don't you think?

**Comment by galaunay** [comment]
It appears that `elpy-autopep8-fix-code` was suffering the same problem:
configurations files in the project root was not taken into account.

It is now fixed.

It also allowed to reduce the complexity of `elpy--fix-code-with-formatter` (less conditionals :) )

File: elpy.el [hunk 1]
```diff
@@ -2099,30 +2099,43 @@ prefix argument is given, prompt for a symbol from the user."
...
+        (directory (or (expand-file-name (elpy-project-root))
+                              default-directory)))
     (if (use-region-p)
-        (let ((new-block (elpy-rpc method (list (elpy-rpc--region-contents))))
-              (beg (region-beginning)) (end (region-end)))
+        (let ((new-block (elpy-rpc method
+                                   (list (elpy-rpc--region-contents)
+                                         directory)))
+              (beg (region-beginning))
+              (end (region-end)))
```
Comment (jorgenschaefer) [inline elpy.el]: Much nicer layout here, thank you. The old one has annoyed me for a while.

File: elpy.el [hunk 2]
```diff
@@ -2099,30 +2099,43 @@ prefix argument is given, prompt for a symbol from the user."
...
            beg end
            (replace-regexp-in-string "\n$" "" new-block))
           (goto-char end)
           (deactivate-mark))
       ;; Vector instead of list, json.el in Emacs 24.3 and before
       ;; breaks for single-element lists of alists.
-      (let ((new-block (elpy-rpc method (vector (elpy-rpc--buffer-contents))))
+      (let ((new-block (elpy-rpc method
+                                 (list (elpy-rpc--buffer-contents)
+                                       directory)))
```
Comment (jorgenschaefer) [inline elpy.el]: The comment can go away, right?

File: elpy/auto_pep8.py [hunk 3]
```diff
@@ -11,11 +12,15 @@
...
 
     """
     if not autopep8:
         raise Fault('autopep8 not installed, cannot fix code.',
                     code=400)
-    return autopep8.fix_code(code, apply_config=True)
+    old_dir = os.getcwd()
+    os.chdir(directory)
+    ret = autopep8.fix_code(code, apply_config=True)
+    os.chdir(old_dir)
```
Comment (jorgenschaefer) [inline elpy/auto_pep8.py]: Use `try`/`finally` here to ensure the directory is changed back even in case of an error.

File: elpy.el (hunk 2 above)
Comment (galaunay) [reply elpy.el]: didn't noticed it,
It is removed now.

File: elpy/auto_pep8.py (hunk 3 above)
Comment (galaunay) [reply elpy/auto_pep8.py]: Done

**Comment by jorgenschaefer** [comment]
Sans the broken tests, this looks good to me! Thank you!

**Comment by galaunay** [comment]
Tests are up to date,
merging !

**Pull request merged**